Tutorials/Tutorial27_PostProcessing/src/Tutorial27_PostProcessing.cpp

```diff
@@ -0,0 +1,615 @@
+        Desc.Format    = TEX_FORMAT_R11G11B10_FLOAT;
+        Desc.BindFlags = BIND_SHADER_RESOURCE | BIND_RENDER_TARGET;
+        m_Resources.Insert(TextureIdx, Device.CreateTexture(Desc));
+    }
+}
+
+void Tutorial27_PostProcessing::GenerateGeometry()
+{
+    const SwapChainDesc& SCDesc = m_pSwapChain->GetDesc();
+    m_GBuffer->Resize(m_pDevice, SCDesc.Width, SCDesc.Height);
+
+    auto& RenderTech = m_RenderTech[RENDER_TECH_GENERATE_GEOMETRY];
+    if (!RenderTech.IsInitialized())
+    {
+        ShaderMacroHelper Macros;
+        Macros.Add("MAX_OBJECT_TRANSFORMS", m_MaxObjectCount);
+
+        const auto VS = CreateShader(m_pDevice, nullptr, "FullScreenTriangleVS.fx", "FullScreenTriangleVS", SHADER_TYPE_VERTEX);
+        const auto PS = CreateShader(m_pDevice, nullptr, "GenerateGeometry.fx", "GenerateGeometryPS", SHADER_TYPE_PIXEL, Macros);
+
+        PipelineResourceLayoutDescX ResourceLayout;
+        ResourceLayout
+            .AddVariable(SHADER_TYPE_PIXEL, "cbCameraAttribs", SHADER_RESOURCE_VARIABLE_TYPE_STATIC)
+            .AddVariable(SHADER_TYPE_PIXEL, "cbObjectTransforms", SHADER_RESOURCE_VARIABLE_TYPE_DYNAMIC);
+
+        RenderTech.InitializePSO(m_pDevice,
+                                 nullptr, "Tutorial27_PostProcessing::GenerateGeometry",
+                                 VS, PS, ResourceLayout,
+                                 {
+                                     m_GBuffer->GetBuffer(GBUFFER_RT_BASE_COLOR)->GetDesc().Format,
+                                     m_GBuffer->GetBuffer(GBUFFER_RT_MATERIAL_DATA)->GetDesc().Format,
+                                     m_GBuffer->GetBuffer(GBUFFER_RT_NORMAL)->GetDesc().Format,
+                                     m_GBuffer->GetBuffer(GBUFFER_RT_OBJECT_IDENTIFIER)->GetDesc().Format,
+                                 },
+                                 m_GBuffer->GetBuffer(GBUFFER_RT_DEPTH)->GetDesc().Format,
+                                 DSS_DepthWrite, BS_Default, false);
+        ShaderResourceVariableX{RenderTech.PSO, SHADER_TYPE_PIXEL, "cbCameraAttribs"}.Set(m_Resources[RESOURCE_IDENTIFIER_CAMERA_CONSTANT_BUFFER]);
+        RenderTech.InitializeSRB(true);
+    }
+
+    const Uint32 CurrFrameIdx = (m_CurrentFrameNumber + 0x0) & 0x1;
+
+    ShaderResourceVariableX{RenderTech.SRB, SHADER_TYPE_PIXEL, "cbObjectTransforms"}.Set(m_Resources[RESOURCE_IDENTIFIER_OBJECT_TRANSFORMS0 + CurrFrameIdx].AsBuffer());
+
+    ScopedDebugGroup DebugGroup{m_pImmediateContext, "GenerateGeometry"};
+
+    m_GBuffer->Bind(m_pImmediateContext, GBUFFER_RT_FLAG_ALL, nullptr, GBUFFER_RT_FLAG_ALL);
+    m_pImmediateContext->SetPipelineState(RenderTech.PSO);
+    m_pImmediateContext->CommitShaderResources(RenderTech.SRB, RESOURCE_STATE_TRANSITION_MODE_TRANSITION);
+    m_pImmediateContext->Draw({3, DRAW_FLAG_VERIFY_ALL});
+    m_pImmediateContext->SetRenderTargets(0, nullptr, nullptr, RESOURCE_STATE_TRANSITION_MODE_NONE);
+}
+
+void Tutorial27_PostProcessing::ComputeMotionVector()
```

Comment:
Why do you need a separate pass for motion vectors?
Why can't they be rendered at the same time with the scene?

Reply:
Let's keep it, as it will allow the user to see how to perform position reconstruction from depth. For OpenGL, this isn't trivial. Also, this might simplify the understanding of the code a bit.